src/azul/indexer/__init__.py

```diff
@@ -45,21 +46,105 @@
+    # their inheritance relationships or how their attributes are annotated
+    # (e.g. specifying `eq=False` has no effect). We want instances of
+    # all subclasses to compare equal as long as `uuid` and `version` are
+    # equal.
+    @final
+    def __eq__(self, other: 'BundleFQID') -> bool:
+        """
+        >>> b1 = BundleFQID(uuid='a', version='b')
+        >>> b2 = BundleFQID(uuid='a', version='b')
+        >>> b1 == b2
+        True
+
+        >>> s1 = SourceRef(id='x', spec=SimpleSourceSpec.parse('y:/0'))
+        >>> sb1 = SourcedBundleFQID(uuid='a', version='b', source=s1)
+        >>> sb2 = SourcedBundleFQID(uuid='a', version='b', source=s1)
+        >>> sb1 == sb2
+        True
+
+        >>> b1 == sb1
+        True
+
+        >>> s2 = SourceRef(id='w', spec=SimpleSourceSpec.parse('z:/0'))
+        >>> sb3 = SourcedBundleFQID(uuid='a', version='b', source=s2)
+        >>> b1 == sb3
+        True
+
+        >>> sb1 == sb3
+        ... # doctest: +NORMALIZE_WHITESPACE
+        Traceback (most recent call last):
+        ...
+        AssertionError: (('a', 'b'),
+        SourceRef(id='x', spec=SimpleSourceSpec(prefix=Prefix(common='', partition=0), name='y')),
+        SourceRef(id='w', spec=SimpleSourceSpec(prefix=Prefix(common='', partition=0), name='z')))
+        """
+        same_bundle = self._nucleus() == other._nucleus()
+        if (
+            same_bundle
+            and isinstance(self, SourcedBundleFQID)
+            and isinstance(other, SourcedBundleFQID)
+        ):
+            assert self.source == other.source, (self._nucleus(), self.source, other.source)
+        return same_bundle
+
+    @final
+    def __hash__(self) -> int:
+        return hash(self._nucleus())
+
+    def __init_subclass__(cls, **kwargs):
+        """
+        >>> @attrs.frozen(kw_only=True)
+        ... class FooBundleFQID(SourcedBundleFQID):
+        ...     foo: str
+        Traceback (most recent call last):
+        ...
+        AssertionError: <class 'azul.indexer.FooBundleFQID'>
+
+        >>> @attrs.frozen(kw_only=True, eq=False)
+        ... class FooBundleFQID(SourcedBundleFQID):
+        ...     foo: str
+        """
+        super().__init_subclass__(**kwargs)
+        assert cls.__eq__ is BundleFQID.__eq__, cls
+        assert cls.__hash__ is BundleFQID.__hash__, cls
+
+    # attrs doesn't allow `order=True` when `eq=False`
+    def __lt__(self, other: 'BundleFQID') -> bool:
+        """
+        >>> list(sorted([
+        ...     BundleFQID(uuid='d', version='e'),
+        ...     BundleFQID(uuid='a', version='c'),
+        ...     BundleFQID(uuid='a', version='b'),
+        ... ]))
+        ... # doctest: +NORMALIZE_WHITESPACE
+        [BundleFQID(uuid='a', version='b'),
+        BundleFQID(uuid='a', version='c'),
+        BundleFQID(uuid='d', version='e')]
+        """
+        return self._nucleus() < other._nucleus()
+
+    def __gt__(self, other: 'BundleFQID'):
```

Comment:
```suggestion
    def __gt__(self, other: Self):
```

Also, consider using https://docs.python.org/3/library/functools.html#functools.total_ordering instead.

It would be good to have doctest cases in `__lt__` that prove that all rich comparisons `< > <= >= = and !=` are implemented and correct.